### ISSUE TITLE
Recommended way to implement undo/redo?

### ISSUE THREAD

**Issue by davidroeca** [jaredpalmer/formik#2339]
## ❓Question

I’m intending on migrating a form previously implemented with [redux-form](https://github.com/redux-form/redux-form) and [redux-undo](https://github.com/omnidan/redux-undo). I currently have a proof of concept implementation of undo/redo using a separate context with useReducer that relies on useFormikContext and operates similarly to redux-undo by storing intermediate values in an object like:

```ts
interface History<Values> {
  past: Array<Values>
  present: Values
  future: Array<Values>
}
```

Undo/redo actions simply read from past/future, setValues(), then update the history.

Is this the only way to achieve undo/redo with formik? Are there other recommended ways? One downside to this approach is that every change (I.e. every keystroke in an input) will be a new part of the history and it’s hard to group them together other than by implementing some minimum time interval between items in the history.

**Comment by malyzeli** [comment]
@davidroeca I'm also in need of undo/redo functionality and I'm going to start working on PoC for our project soon. My plan for implementation was basically the same as yours, since previously I also used `redux-undo`. I have some ideas for "grouping" changes in the history - for example using blur handlers to store single history state per individual field change. I would be glad to cooperate. Maybe @jaredpalmer will give us some ideas as well?

**Comment by johnrom** [comment]
That's a pretty heavy history, I would recommend instead, if it was possible in the future to subscribe to Formik's reducer itself, to build a history of changes instead of maintaining the full value list. There should also be a limit of how long prev and next can get.

Not sure if it can be implemented with the current API, since I believe we can only subscribe using `onChange`s, and not changes using `setFieldValue()`. If you aren't using setFieldValue to do anything like setting dependent values, this might be possible for you.

```tsx
// each prev and next is a "set" of changes, instead of the full values object
export interface FormikHistory<TValues> {
    prev: Array<Array<FormikMessage<TValues>['onChange']>>;
    snapshot: TValues,
    next: Array<Array<FormikMessage<TValues>['onChange']>>; 
}
```

**Comment by malyzeli** [comment]
@johnrom I don't see how storing `FormikMessage`s instead of actual values might result in "lighter" history


```js
const stateSnapshot = { field: "username", value: "johndoe" }

const formikMessageSnapshot = { 
  type: "SET_FIELD_VALUE", 
  payload: { 
    field: "username",
    value: "johndoe"
  }
}
```

**Comment by johnrom** [comment]
With my example above (which I just fixed now that I have Formik's types in front of me, here's a comparison of changes in form state with 5 fields.

```
// typed my name, a space, then hit undo
const heavyHistory = {
    prev: [
        { firstName: '', lastName: '', age: 0, favoriteColor: '', favoriteDinosaur: '' },
        { firstName: 'j', lastName: '', age: 0, favoriteColor: '', favoriteDinosaur: '' },
        { firstName: 'jo', lastName: '', age: 0, favoriteColor: '', favoriteDinosaur: '' },
        { firstName: 'joh', lastName: '', age: 0, favoriteColor: '', favoriteDinosaur: '' },
    ],
    currentState: 
        { firstName: 'john', lastName: '', age: 0, favoriteColor: '', favoriteDinosaur: '' },
    next: [
        { firstName: 'john ', lastName: '', age: 0, favoriteColor: '', favoriteDinosaur: '' },
    ]
}
const lightHistory = {
    prev: [
        { fieldName: 'firstName', value: '' },
        { fieldName: 'firstName', value: 'j' },
        { fieldName: 'firstName', value: 'jo' },
        { fieldName: 'firstName', value: 'joh' },
    ],
    currentState: 
        { firstName: 'john', lastName: '', age: 0, favoriteColor: '', favoriteDinosaur: '' },
    next: [
        { fieldName: 'firstName', value: 'john ' },
    ]
}
```

**Comment by malyzeli** [comment]
@johnrom Yep, totally agree with that - I just didn't noticed OP was talking about storing all values. 

What is the reasoning for storing full form data in `currentState`, considering I can read the current state from Formik context anytime (and store only the changes in my history stack)?

**Comment by davidroeca** [comment]
Unfortunately my form is not so simple and I'd need to track changes made by the user, or those made by automatic updates to dependent fields (the dropdowns in my form have options that depend on one another). So my only option as it stands is to store all values unless formik exposes some of its internals.

**Comment by davidroeca** [comment]
This can be limited by what I previously mentioned (set a minimum time interval e.g. 1-2 seconds that limits updates to the history) and by limiting the length of past/future, so that the history doesn't balloon.

**Comment by johnrom** [comment]
I was just using OP's example where he was managing state higher in the hierarchy, which is not ideal due to constant re-rendering of Formik. Instead I propose allowing a dev to override the reducer itself with a redux-like version that supports subscription so these two can coexist at the same level and not depend on keeping context in sync.

**Comment by johnrom** [comment]
Being able to override and subscribe to the reducer would solve all sorts of issues like #1218 #172 

Related issue: #1854

**Comment by malyzeli** [comment]
Is anything wrong with approach of this kind? 

```jsx
function useFormikHistory(props) {
  const formik = useFormik(props);
  
  const history = useState();
  // or useReducer, whatever

  function setValues(values, /*...*/) {
    // store whatever you need then pass values to formik
    formik.setValues(values, /*...*/);
  }
  
  function setFieldValue(/*...*/) {
    // same same, but different
  }
 
  return { 
    ...formik,
    setFieldValue,
    setValues,
  };
}
```

Seems cleaner to me than messing with Formik's internals...

Then of course you need to add some handler for undo/redo logic, which would internally use Formik's `setValues`/`setFieldValue` again.

I will share my results if I manage to create something usable.

**Comment by davidroeca** [comment]
@johnrom agreed, these issues seem pretty relevant. #1633 might also help.

@malyzeli That might solve it, but then you wouldn't have access to `useField`, `useFormikContext`, etc. and would have to write your own. Not that's necessarily a bad thing, but you wouldn't really be using the full extent of what formik has to offer and would have to write your own versions of each. Maybe that's what we're stuck with here.

**Comment by malyzeli** [comment]
@davidroeca not if you pass value of this `useFormikHistory` to `FormikProvider`, which is obviously the next step. Then you should be able to use all the stuff as usual - from child components point of view nothing really changes.

See how [`Formik`](https://github.com/jaredpalmer/formik/blob/26c4f8627a5ecfd81ec2196c7a9687b3f39f2836/packages/formik/src/Formik.tsx#L997) component is implemented, ultimately it can be simplified pretty much to this:

```jsx
export function Formik(props) {
  const formik = useFormik(props);
  return (
    <FormikProvider value={formik}>
      {children}
    </FormikProvider>
  );
}
```

Which means you can do this:

```jsx
export function FormikHistory(props) {
  const formik = useFormikHistory(props);
  /* might put your event handlers for undo/redo here? */
  return (
    <FormikProvider value={formik}>
      {children}
    </FormikProvider>
  );
}
```

And you are back in business...

**Comment by davidroeca** [comment]
:thinking: interesting... the main downside I see with this approach that I see is that none of this is well-documented so it's unclear whether some of these components will change functionality or be removed in the future.

Regardless, it seems promising so I may give this a shot.

**Comment by johnrom** [comment]
yeah that's the best way to get started right now pending Jared's feedback on passing in a custom reducer. @davidroeca the best way to get it well-documented would be to open source your wrapper and document it there :)

**Comment by malyzeli** [comment]
@davidroeca well, change is the only constant - you might have to update your code in future, but that will probably happen regardless on chosen solution. :smile:

I prefer enhancing tools/libraries via this kind of "wrapping", it is definitely more flexible than digging into any internals, and also it's exactly what React hooks are good for!

If you put it into separate repo, feel free to invite me for collaboration, I would be glad to help.

**Comment by davidroeca** [comment]
I'll play around with this implementation a bit with my form and will check back in here when I feel like it's in a good place to open source, hopefully sometime this week :crossed_fingers:

**Comment by davidroeca** [comment]
@malyzeli I'm not sure how much progress you've made here but I've hit a bit of a snag with how `useFormik` is implemented.

It mostly has to do with the fact that certain internal methods rely on `setFieldValue` and there is a dependency hierarchy. At a shallow level, the following methods are derived from `setFieldValue`:
* `executeChange`
* `imperativeMethods`
* `getFieldHelpers`

Only `imperativeMethods` relies on `setValues`.

Following on down this dependency hierarchy, even more methods/attributes rely on these. Since this all happens in one go, changing `setFieldValue` after the fact won't allow us to achieve the desired functionality (at least for my case, where I use a bunch of `useField`s). Let me know if you've found a work-around.

**Comment by malyzeli** [comment]
@davidroeca hmm good point, I didn't realise `useField` is NOT using `setFieldValue` from context but it's wired internally - was thinking just about setting values imperatively. 

I will look into it more next week, now I have some other issues to solve first.

**Comment by OoDeLally** [comment]
I've just created a package for that. It uses Typescript and React Hooks (no redux, sorry).
https://github.com/OoDeLally/formik-undo
Feel free to use it and submit bugs and PRs.
